allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -271,6 +318,165 @@ def forward(self,  # type: ignore
+                entity_neighbors = [entity2index[n] for n in entity2neighbors[entity]]
+                # Pad with -1 instead of 0, since 0 represents a neighbor index.
+                padded = pad_sequence_to_length(entity_neighbors, num_neighbors, lambda: -1)
+                neighbor_indexes.append(padded)
+            neighbor_indexes = pad_sequence_to_length(neighbor_indexes,
+                                                      num_entities,
+                                                      lambda: [-1] * num_neighbors)
+            batch_neighbors.append(neighbor_indexes)
+        return Variable(tensor.data.new(batch_neighbors)).long()
+
+    @staticmethod
+    def _get_type_vector(worlds: List[WikiTablesWorld],
+                         num_entities: int,
+                         tensor: Variable) -> Tuple[torch.LongTensor, Dict[int, int]]:
+        """
+        Produces the one hot encoding for each entity's type. In addition,
+        a map from a flattened entity index to type is returned to combine
+        entity type operations into one method.
+
+        Parameters
+        ----------
+        worlds : ``List[WikiTablesWorld]``
+        num_entities : ``int``
+        tensor : ``torch.Tensor``
+            Used for copying the constructed list onto the right device.
+
+        Returns
+        -------
+        A ``torch.LongTensor`` with shape ``(batch_size, num_entities, num_types)``.
+        entity_types : ``Dict[int, int]``
+            This is a mapping from ((batch_index * num_entities) + entity_index) to entity type id.
+        """
+        entity_types = {}
+        batch_types = []
+        for batch_index, world in enumerate(worlds):
+            types = []
+            for entity_index, entity in enumerate(world.table_graph.entities):
+                cell_type_one_hot = [1, 0]
+                row_type_one_hot = [0, 1]
+                entity_type = 0 if entity.startswith('fb:cell') else 1
+                types.append((cell_type_one_hot if entity_type == 0 else row_type_one_hot))
+
+                # For easier lookups later, we're actually using a _flattened_ version
+                # of (batch_index, entity_index) for the key, because this is how the
+                # linking scores are stored.
+                flattened_entity_index = batch_index * num_entities + entity_index
+                entity_types[flattened_entity_index] = entity_type
+            padded = pad_sequence_to_length(types, num_entities, lambda: [0, 0])
+            batch_types.append(padded)
+        return Variable(tensor.data.new(batch_types)), entity_types
+
+    @staticmethod
+    def _get_linking_probabilities(worlds: List[WikiTablesWorld],
+                                   linking_scores: Variable,
+                                   question_mask: Variable,
+                                   entity_type_dict: Dict[int, int],
+                                   tensor: Variable) -> torch.FloatTensor:
```

Comment:
You don't actually need to take this tensor, because you've already got `linking_scores` that you can copy from instead.  We only needed it in `_get_neighbor_indices` because there wasn't any other tensor to copy from.